daemon/src/model/cfd.rs

```diff
@@ -2389,6 +2390,135 @@ mod tests {
+        let leverage = Leverage::new(2).unwrap();
+        let opening_price = Price::new(dec!(10000)).unwrap();
+
+        let taker_keys = crate::keypair::new(&mut rand::thread_rng());
+        let maker_keys = crate::keypair::new(&mut rand::thread_rng());
+
+        let cfd = Cfd::taker_long()
+            .with_quantity(quantity)
+            .with_opening_price(opening_price)
+            .with_leverage(leverage)
+            .dummy_open(dummy_event_id());
+
+        let _ = assert!(cfd.start_rollover().is_ok());
+
+        let (cfd, _, _, _) = cfd
+            .with_lock(taker_keys, maker_keys)
+            .dummy_collab_settlement_taker(opening_price);
+
+        let err = cfd.start_rollover().unwrap_err();
+        let no_rollover_reason = err.downcast::<NoRolloverReason>().unwrap();
+        assert_eq!(no_rollover_reason, NoRolloverReason::Closed);
+    }
+
+    #[test]
+    fn given_ongoing_collab_settlement_then_cannot_start_rollover() {
+        let quantity = Usd::new(dec!(10));
+        let leverage = Leverage::new(2).unwrap();
+        let opening_price = Price::new(dec!(10000)).unwrap();
+
+        let cfd = Cfd::taker_long()
+            .with_quantity(quantity)
+            .with_opening_price(opening_price)
+            .with_leverage(leverage)
+            .dummy_open(dummy_event_id());
+
+        let _ = assert!(cfd.start_rollover().is_ok());
+
+        let cfd = cfd.dummy_ongoing_collab_settlement();
+
+        let err = cfd.start_rollover().unwrap_err();
+        let no_rollover_reason = err.downcast::<NoRolloverReason>().unwrap();
+        assert_eq!(
+            no_rollover_reason,
+            NoRolloverReason::InCollaborativeSettlement
+        );
+
+        let cfd = Cfd::maker_short()
+            .with_quantity(quantity)
+            .with_opening_price(opening_price)
+            .with_leverage(leverage)
+            .dummy_open(dummy_event_id());
+
+        let _ = assert!(cfd.start_rollover().is_ok());
+
+        let cfd = cfd.dummy_ongoing_collab_settlement();
+
+        let err = cfd.start_rollover().unwrap_err();
+        let no_rollover_reason = err.downcast::<NoRolloverReason>().unwrap();
+        assert_eq!(
+            no_rollover_reason,
+            NoRolloverReason::InCollaborativeSettlement
+        );
+    }
+
+    #[test]
+    fn given_ongoing_rollover_then_can_start_collaborative_settlement() {
+        let quantity = Usd::new(dec!(10));
+        let leverage = Leverage::new(2).unwrap();
+        let opening_price = Price::new(dec!(10000)).unwrap();
+        let order_id = OrderId::default();
+
+        let taker_keys = crate::keypair::new(&mut rand::thread_rng());
+        let maker_keys = crate::keypair::new(&mut rand::thread_rng());
+
+        let taker_long = Cfd::taker_long()
+            .with_id(order_id)
+            .with_quantity(quantity)
+            .with_opening_price(opening_price)
+            .with_leverage(leverage)
+            .dummy_open(dummy_event_id());
```

Comment:
this boilerplate makes these tests not that readable.... I'll post a quick PR on top of it if you don't mind, I have something in mind for it 😀

Reply:
definitely! If you have something in mind already that's very much appreciated. I wanted to get them done with what's currently there and then do a round of refactoring myself. If you already start / do with that's even better :)